syntax = "proto3";
package replica;


service Replication {
    rpc RegisterNode (NodeAddress) returns (NodeID) {}
    rpc GetNodeAddress (NodeID) returns (NodeAddress) {}
    rpc Heartbeat (PingMessage) returns (ViewState) {}
}

service PrimaryBackup {
    rpc StateUpdateStream (Empty) returns (stream StateUpdate) {}
}

message Empty {}

message NodeAddress {
    string address = 1;
    string port = 2;
}

message NodeID {
    string val = 1;
}

message PingMessage {
    string node_id = 1;
}

message ViewState {
    string primary_id = 1;
    string backup_id = 2;
    string view_id = 3;
}

message StateUpdate {
    bytes state = 1;
}